### PR TITLE
IFF refactor

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Refactor IFF to be using factions rather than access.
Also allow to have ammo enabling IFF on guns without IFF 

## Why It's Good For The Game

Better code, also will be good for HvH

## Changelog
:cl:
refactor: Refactor IFF
/:cl:

### CODE REVIEW

File: code/modules/reqs/supplypacks.dm [hunk 1]
```diff
@@ -616,7 +616,7 @@ AMMO
 /datum/supply_packs/ammo/back_fuel_tank_x
 	name = "Type X back fuel tank"
 	contains = list(/obj/item/ammo_magazine/flamer_tank/backtank/X)
-	cost = 100
+	cost = 60
```
Comment (BraveMole) [inline code/modules/reqs/supplypacks.dm:619]:
```suggestion
	cost = 100
```

**Comment by ilikesaying** [comment]
>Also allow to have ammo enabling IFF on guns without IFF

IFF GPMG meme

File: code/game/objects/machinery/sentries.dm [hunk 2]
```diff
@@ -228,7 +228,8 @@
 	req_one_access = list(ACCESS_MARINE_ENGINEERING, ACCESS_MARINE_ENGPREP, ACCESS_MARINE_LEADER)
 	hud_possible = list(MACHINE_HEALTH_HUD, SENTRY_AMMO_HUD)
 	var/turret_flags = TURRET_HAS_CAMERA|TURRET_SAFETY|TURRET_ALERTS
-	var/list/iff_signal = list(ACCESS_IFF_MARINE)
+	/// The faction used to determine friendlies from hostiles
+	var/faction
```
Comment (SplinterGP) [inline code/game/objects/machinery/sentries.dm:232]: I prefer this to be like it was in the comments on the iff signal, a bitfiel so you can have multiple iff signals

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/sentries.dm (hunk 2 above)
Comment (BraveMole) [reply code/game/objects/machinery/sentries.dm:232]: i mean this was never used, and i don't think anyone plan to use it?

I certainly don't need it for hvh mode

Comment (SplinterGP) [reply code/game/objects/machinery/sentries.dm:232]: > 
> 
> i mean this was never used, and i don't think anyone plan to use it? And you mean a list, access are not bitfields
> 
> I certainly don't need it for hvh mode

the comment said it was to be turned into a bitfield later

File: code/datums/jobs/job/pmc.dm [hunk 3]
```diff
@@ -95,7 +95,7 @@
 
 	H.equip_to_slot_or_del(new /obj/item/reagent_containers/hypospray/autoinjector/oxycodone, SLOT_IN_R_POUCH)
 	H.equip_to_slot_or_del(new /obj/item/reagent_containers/hypospray/autoinjector/oxycodone, SLOT_IN_R_POUCH)
-	H.equip_to_slot_or_del(new /obj/item/explosive/mine/pmc, SLOT_IN_R_POUCH)
```
Comment (TiviPlus) [inline code/datums/jobs/job/pmc.dm:98]: ?????

File: code/game/objects/items/explosives/mine.dm [hunk 4]
```diff
@@ -16,8 +16,8 @@ Stepping directly on the mine will also blow it up
 	throw_speed = 3
 	flags_atom = CONDUCT
 
-	/// IFF signal - used to determine friendly units
-	var/list/iff_signal = list(ACCESS_IFF_MARINE)
+	/// Used to determine friendly units
+	var/faction
```
Comment (TiviPlus) [inline code/game/objects/items/explosives/mine.dm:20]: this should be bitflags

File: code/game/objects/machinery/sentries.dm (hunk 2 above)
Comment (TiviPlus) [inline code/game/objects/machinery/sentries.dm:232]: also

File: code/modules/projectiles/gun_system.dm [hunk 5]
```diff
@@ -915,12 +913,13 @@ and you're good to go.
...
 
-/obj/item/weapon/gun/proc/apply_gun_modifiers(obj/projectile/projectile_to_fire, atom/target)
+/obj/item/weapon/gun/proc/apply_gun_modifiers(obj/projectile/projectile_to_fire, atom/target, mob/living/user)
 	projectile_to_fire.shot_from = src
 	projectile_to_fire.damage *= damage_mult
 	projectile_to_fire.damage_falloff *= damage_falloff_mult
 	projectile_to_fire.projectile_speed += shell_speed_mod
-	projectile_to_fire.projectile_iff = gun_iff_signal
+	if(flags_gun_features & GUN_IFF || flags_gun_features & GUN_IS_AIMING|| projectile_to_fire.ammo.flags_ammo_behavior & AMMO_IFF)
+		projectile_to_fire.factions = list(user.faction)
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:922]: is it a list or not you treat is as both

File: code/modules/projectiles/projectile.dm [hunk 6]
```diff
@@ -67,8 +67,8 @@
 
 	var/scatter = 0 //Chance of scattering, also maximum amount scattered. High variance.
 
-	/// Used to transfer iff_signal from source to projectile.
-	var/list/projectile_iff = null
+	/// The factions this projectile will ignore, used for iff
+	var/factions = list()
```
Comment (TiviPlus) [inline code/modules/projectiles/projectile.dm:71]: ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by SplinterGP** [comment]
reopening at request

File: code/datums/jobs/job/deathsquad.dm [hunk 7]
```diff
@@ -187,7 +187,7 @@
 	mask = /obj/item/clothing/mask/gas/PMC
 	head = /obj/item/clothing/head/helmet/marine/veteran/PMC/commando
 	glasses = /obj/item/clothing/glasses/night/m56_goggles
-	suit_store = /obj/item/weapon/gun/rifle/standard_smartmachinegun/ds
+	suit_store = /obj/item/weapon/gun/rifle/standard_smartmachinegun
```
Comment (TiviPlus) [inline code/datums/jobs/job/deathsquad.dm:190]: unrelated change

**Review by TiviPlus — APPROVED** [review]

File: code/datums/jobs/job/deathsquad.dm (hunk 7 above)
Comment (BraveMole) [reply code/datums/jobs/job/deathsquad.dm:190]: /obj/item/weapon/gun/rifle/standard_smartmachinegun/ds is deleted

**Review by TiviPlus — APPROVED** [review]
